subsys/net/lib/websocket/websocket.c

```diff
@@ -818,178 +863,103 @@ int websocket_recv_msg(int ws_sock, uint8_t *buf, size_t buf_len,
 	}
 #endif /* CONFIG_NET_TEST */
 
-	/* If we have not received the websocket header yet, read it first */
-	if (!ctx->header_received) {
-#if defined(CONFIG_NET_TEST)
-		size_t input_len = MIN(ctx->tmp_buf_len - ctx->tmp_buf_pos,
-				       test_data->input_len);
+	ctx->data_buf_count = 0;
+	ctx->data_buf_size = buf_len;
+	ctx->data_buf = buf;
+
+	int exit_from_loop = false;
 
-		memcpy(&ctx->tmp_buf[ctx->tmp_buf_pos], test_data->input_buf,
-		       input_len);
-		test_data->input_buf += input_len;
-		ret = input_len;
+	do {
+		if (ctx->tmp_buf_pos == 0) {
+#if defined(CONFIG_NET_TEST)
+			size_t input_len = MIN(ctx->tmp_buf_len,
+					test_data->input_len - test_data->input_pos);
+
+			if (input_len > 0) {
+				memcpy(ctx->tmp_buf,
+					&test_data->input_buf[test_data->input_pos], input_len);
+				test_data->input_pos += input_len;
+				ret = input_len;
+			} else {
+				/* emulate timeout */
+				errno = EAGAIN;
+				ret = -1;
+			}
 #else
-		ret = recv(ctx->real_sock, &ctx->tmp_buf[ctx->tmp_buf_pos],
-			   ctx->tmp_buf_len - ctx->tmp_buf_pos,
-			   K_TIMEOUT_EQ(tout, K_NO_WAIT) ? MSG_DONTWAIT : 0);
+			ret = recv(ctx->real_sock, ctx->tmp_buf, ctx->tmp_buf_len,
+					K_TIMEOUT_EQ(tout, K_NO_WAIT) ? MSG_DONTWAIT : 0);
 #endif /* CONFIG_NET_TEST */
 
-		if (ret < 0) {
-			return -errno;
-		}
-
-		if (ret == 0) {
-			/* Socket closed */
-			return 0;
-		}
-
-		ctx->tmp_buf_pos += ret;
-
-		if (ctx->tmp_buf_pos >= MIN_HEADER_LEN) {
-			bool masked;
-
-			/* Now we will be able to figure out what is the
-			 * actual size of the header.
-			 */
-			if (websocket_parse_header(&ctx->tmp_buf[0],
-						   ctx->tmp_buf_pos,
-						   &masked,
-						   &ctx->masking_value,
-						   &ctx->message_len,
-						   &ctx->message_type,
-						   &header_len)) {
-				ctx->masked = masked;
-
-				if (message_type) {
-					*message_type = ctx->message_type;
+			if (ret < 0) {
+				ret = -errno;
+				if ((ret == -EAGAIN) && (ctx->data_buf_count > 0)) {
+					/* go to unmasking */
+					break;
 				}
-			} else {
-				return -EAGAIN;
+				return ret;
 			}
-		} else {
-			return -EAGAIN;
-		}
-
-		if (ctx->tmp_buf_pos < header_len) {
-			return -EAGAIN;
-		}
 
-		/* All of the header is now received, we can read the payload
-		 * data next.
-		 */
-		ctx->header_received = true;
-
-		if (HEXDUMP_RECV_PACKETS) {
-			LOG_HEXDUMP_DBG(&ctx->tmp_buf[0], header_len,
-					"Header");
-			NET_DBG("[%p] masked %d mask 0x%04x hdr %zd msg %zd",
-				ctx, ctx->masked,
-				ctx->masked ? ctx->masking_value : 0,
-				header_len, (size_t)ctx->message_len);
-		}
-
-		ctx->total_read = 0;
+			if (ret == 0) {
+				/* Socket closed */
+				return -ENOTCONN;
+			}
 
-		memmove(ctx->tmp_buf, &ctx->tmp_buf[header_len],
-			ctx->tmp_buf_len - header_len);
-		ctx->tmp_buf_pos -= header_len;
+			ctx->tmp_buf_pos = ret;
```

Comment:
I think this variable name needs an update - it no longer represents currently processed data **position** but rather the length of the data within `tmp_buf`.

Reply:
Rename `tmp_buf` to `recv_buf` and use `struct websocket_buffer` for `recv_buf` and `payload`.
```c
/**
 * Description of external buffers for payload and receiving
 */
struct websocket_buffer {
	/* external buffer */
	uint8_t *buf;
	/* size of external buffer */
	size_t size;
	/* data length in external buffer */
	size_t count;
}; 
```